### PR TITLE
Dynamic jobs

### PR DESCRIPTION
**Pull request by linkdotnet**
## Pull request description
Give the developer means to add and remove jobs dynamically.

### PR meta checklist
- [x] Pull request is targeted at `main` branch for code   
- [x] Pull request is linked to all related issues, if any.

### Code PR specific checklist
- [x] My code follows the code style of this project and AspNetCore coding guidelines.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] I have updated the appropriate sub section in the _CHANGELOG.md_.
- [x] I have added, updated or removed tests to according to my changes.
  - [x] All tests passed.

### CODE REVIEW

**Comment by falvarez1** [comment]
@linkdotnet 

I appreciate the effort to provide a way to disable a CRON job. However, I don't think using an arbitrary CRON expression like "31st of February" is the right approach. Here are my concerns and suggestions:

1. **Misleading Intent**: Using an unattainable date in the CRON expression to disable a job can be misleading and confusing for users. It doesn't clearly convey the intent to disable the job, and users may not understand why this specific date is chosen.
2. **Potential Issues**: Relying on an invalid date could lead to unexpected behavior or bugs in the future, especially if the underlying CRON library updates its validation rules.
3. **Proper Solution**: We should implement a dedicated feature to disable jobs. This would make the intent clear and the code more maintainable. Here’s a suggestion on how we could approach it:

**Add a Disable Method**: Add a method in the `IRuntimeJobRegistry` interface to disable jobs by name.
**Update Endpoint**: Update the endpoint to use this new method.

```csharp
public interface IRuntimeJobRegistry
{
    ...
    void DisableJob(string name);
}
```
```csharp
app.MapPut("/disable-job", (IRuntimeJobRegistry registry) => 
{
    registry.DisableJob("MyName");
    return TypedResults.Ok();
});
```
We can use the February 31st CRON expression as a temporary internal solution, allowing us to evolve the feature later without changing the API. By following this approach, we make it explicit that a job is being disabled, which enhances code readability and maintainability. This ensures users have a clear and intuitive way to disable jobs, aligning with the _principle of least astonishment_.

**Comment by linkdotnet** [comment]
Yeah that is a good idea. Probably for now I just remove that thing entirely from the docs!
We can add `Disable` / `Enable` later on, if someone really needs that.

**Comment by linkdotnet** [comment]
ToDo:
- [ ] `GetAllRecurringJobs`

Currenly I have the following API:
```csharp
IReadOnlyCollection<RecurringJobSchedule> GetAllRecurringJobs();

public sealed record RecurringJobSchedule(Type JobType, string? JobName, string? CronExpression, TimeZoneInfo TimeZone);
```

@falvarez1 I am fighting to see if I like the name and if the information given is good enough. I would like to expose as least as possible.

**Comment by falvarez1** [comment]
If you want we can just add a single method like `void SetJobEnabled(string name, bool isEnabled);` and manage it initially using this Feb 31st approach, then improve it in the state changes branch.

**Comment by linkdotnet** [comment]
> If you want we can just add a single method like `void SetJobEnabled(string name, bool isEnabled);` and manage it initially using this Feb 31st approach, then improve it in the state changes branch.

Yeah good idea! Still thinking of tackling that in another PR. I still have some open thoughts, for example: Would a user like to create a scheduled job that is directly disabled?

**Comment by falvarez1** [comment]
> Would a user like to create a scheduled job that is directly disabled?

This seems like an edge case to me. Users can always create a job and then immediately disable it, assuming we expose the ability to disable jobs. This could be useful if they're creating a template and want to clone it later, but that's beyond our current scope.

**Comment by linkdotnet** [comment]
I guess if we merge this before #85 then you will have a quite some merge conflicts. @falvarez1

File: src/NCronJob/Registry/JobRegistry.cs [hunk 1]
```diff
@@ -14,20 +14,78 @@ public JobRegistry(IEnumerable<JobDefinition> jobs)
...
+    public JobDefinition? FindJobDefinition(string jobName)
+        => allJob.Find(j => j.CustomName == jobName);
+
+    public JobDefinition GetJobDefinitionForInstantJob<T>()
+    {
+        var definition = allJob.First(j => j.Type == typeof(T));
+        if (definition.CronExpression is not null)
+        {
+            var instantDefinition = definition.CreateInstantVersion();
+            allJob.Add(instantDefinition);
```
[the remark this answers is not in the document]
Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:36]: Yeah that is a very fair point! Will do so

File: src/NCronJob/Registry/JobDefinition.cs [hunk 2]
```diff
@@ -14,6 +14,16 @@ internal sealed record JobDefinition(
 
     public string JobName { get; } = JobName ?? Type.Name;
 
+    public string? CustomName { get; set; }
+
+    public CronExpression? CronExpression { get; set; } = CronExpression;
+
+    public string? CronExpressionString { get; set; }
```
[the remark this answers is not in the document]
Comment (linkdotnet) [reply src/NCronJob/Registry/JobDefinition.cs:21]: Exactly that is the goal. The issue with `CronExpression?.ToString` is two-folded:
```csharp
var cron = CronExpression.Parse("*  * * * *"); // Extra whitespace
cron.ToString(); // Is * * * * * without the extra whitespace

var cron = CronExpression.Parse("*/2 * * * *");
cron.ToString(); // Is 0,2,4,6,8,10,... * * * *
```

So they are very much an implementation detail - that said, I will rename the property to reflect that.

File: src/NCronJob/Registry/JobRegistry.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs]: Done

File: src/NCronJob/Registry/JobDefinition.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (linkdotnet) [reply src/NCronJob/Registry/JobDefinition.cs]: Renamed and added explanation why that extra prop

**Comment by falvarez1** [comment]
@linkdotnet the PR looks good to me. Let me know if you want my help with resolving merge conflicts.

**Comment by linkdotnet** [comment]
Yeah I tried and there are many.
Therefore I started from scratch: https://github.com/NCronJob-Dev/NCronJob/tree/dynamic-jobs-remastered

but here are a few things missing.

**Comment by linkdotnet** [comment]
@falvarez1 I dumped that into a chat message instead of posting here.

**Comment by falvarez1** [comment]
Ok. I'm assuming you'll be re-creating a new PR as well. If you create it as a draft I can review as it's being worked on.

**Pull request closed without merging**